src/Myriad.Core/Ast.fs

```diff
@@ -251,27 +330,49 @@ module Ast =
             let expr = defaultArg expr (SynExpr.CreateTyped(SynExpr.CreateNull, SynType.CreateUnit))
-            let bind = DebugPointForBinding.NoDebugPointAtInvisibleBinding
-            SynBinding.Binding(access, SynBindingKind.NormalBinding, isInline, isMutable, attributes, xmldoc, valData, headPat, returnInfo, expr, range0, bind )
+            let bind = DebugPointAtBinding.NoneAtLet
+            let trivia = { LetKeyword = Some range0
+                           EqualsRange = Some range0 }
+            SynBinding.SynBinding(access, SynBindingKind.Normal, isInline, isMutable, attributes, xmldoc, valData, headPat, returnInfo, expr, range0, bind, trivia)
             
             
     type SynModuleDecl with
         static member CreateLet(bindings, ?isRecursive) =
             let isRecursive = defaultArg isRecursive false
             SynModuleDecl.Let(isRecursive, bindings, range0)
             
+        static member CreateNestedModule(ci, decls, ?isRec, ?isCont) =
+            let isRec = defaultArg isRec false
+            let isCont = defaultArg isCont false
+            let trivia = {SynModuleDeclNestedModuleTrivia.EqualsRange = Some range0; ModuleKeyword = Some range0 }
+            SynModuleDecl.NestedModule(ci, isRec, decls, isCont, range0, trivia)
+            
     type SynBindingReturnInfo with
         static member Create(typeName, ?attributes) =
             let attributes = defaultArg attributes SynAttributes.Empty
             SynBindingReturnInfo.SynBindingReturnInfo(typeName, range0, attributes)
             
     type SynUnionCase with
         member x.HasFields =
-            let (SynUnionCase.UnionCase(_,_,typ,_,_,_)) = x
+            let (SynUnionCase.SynUnionCase(_,_,typ,_,_,_,_)) = x
             match typ with
-            | UnionCaseFields cases -> not cases.IsEmpty
+            | SynUnionCaseKind.Fields fields -> not fields.IsEmpty
             | _ -> false
             
     type SynMatchClause with
         static member Create(pat, whenExp, result) =
-            SynMatchClause.Clause(pat, whenExp, result, range0, DebugPointForTarget.No)
+            let trivia = {SynMatchClauseTrivia.ArrowRange = Some range0; BarRange = Some range0}
+            SynMatchClause.SynMatchClause(pat, whenExp, result, range0, DebugPointAtTarget.No, trivia)
+
+    open DynamicReflection
+    type SynExpr with
+        static member CreateLambda(pats: SynPat list, body: SynExpr, ?isMember: bool) =
+            let isMember = defaultArg isMember false
+            let compiler = System.Reflection.Assembly.Load("FSharp.Compiler.Service")
+            let syntaxTreeOps = compiler.GetType("FSharp.Compiler.SyntaxTreeOps")
+            let synArgNameGenerator = compiler.GetType("FSharp.Compiler.SyntaxTreeOps+SynArgNameGenerator")
+            let nameGen = synArgNameGenerator?``.ctor``()
+            syntaxTreeOps?mkSynFunMatchLambdas(nameGen, isMember, range0, pats, Some range0, body)
```

Comment:
Loooool this is amazing :D

Might be worth it to cover this with a test that breaks if the shape changes.

Reply:
Basically myriad fails to run any tests is the function changes , the lenses plugin calls this so it fails with a nice runtime error :-)

I raised a PR for the function `mkSynFunMatchLambdas` to be surfaced in FCS.

Reply:
They are originally from FsAst but with fixes and additions that Ive found, I felt all the To/FromRcd were awkward so culled them and added a host of others, because I have quick in myriad I moved them over and added a bunch more.